packages/use-sync-external-store/src/useSyncExternalStore.js

```diff
@@ -7,6 +7,182 @@
+      // Using a layout effect to update the ref wouldn't solve this because it
+      // wouldn't account for layout effects that run earlier within the same
+      // commit phase.
+      //
+      // To solve, we null out getSnapshot *in the render phase* whenever it
+      // changes, then set it again in the commit phase. That will prevent us
+      // from using it to bail out if we receive an update during that time.
+      inst.getSnapshot = null;
+    }
+  }
+
+  // Track getSnapshot with a passive effect. It's tempting to make this a
+  // layout effect, but as the previous comment explains, that alone wouldn't
+  // prevent us from using a stale getSnapshot. Since we use a different
+  // strategy to account for that problem, a passive effect is sufficient.
+  useEffect(() => {
+    inst.getSnapshot = getSnapshot;
+  }, [getSnapshot]);
+
+  // Subscribe in a layout effect to avoid a flicker. Resubscribe only when
+  // subscribe or getSnapshot changes.
+  // TODO: Circumvent SSR warning
+  useLayoutEffect(() => {
+    const handleStoreChange = () => {
+      // TODO: Because there is no cross-renderer API for batching updates, it's
+      // up to the consumer of this library to wrap their subscription event
+      // with unstable_batchedUpdates. Should we try to detect when this isn't
+      // the case and print a warning in development?
+
+      // Track the current state of the store. We track this within the
+      // subscription handler instead of an effect to account for updates that
+      // happen in between render and commit -- a similar scenario to the one
+      // described above regarding getSnapshot. So this represents the most
+      // recent store state, not the most recent React state.
+
+      // If getSnapshot was changed during render and we haven't reached the
+      // commit effect yet, then it will be null. Otherwise, we can use it to
+      // bail out early.
+      const latestGetSnapshot = inst.getSnapshot;
+      if (latestGetSnapshot !== null) {
+        const prevValue = inst.value;
+        try {
+          const nextValue = getSnapshot();
+          if (!is(prevValue, nextValue)) {
+            // Force a re-render.
+            inst.value = nextValue;
+            setVersion(bumpVersion);
+          }
+        } catch (error) {
+          // Force a re-render.
+          setVersion(bumpVersion);
+        }
+      } else {
+        // Force a re-render.
+        setVersion(bumpVersion);
+      }
+    };
+
+    // When we initially subscribe, we need to check if there was a mutation in
+    // between render and commit. If there was, we may need to re-render.
+    // Subsequent changes will be detected in the subscription handler.
+    const valueDuringRender = value;
+    try {
+      const nextValue = getSnapshot();
+      if (!is(valueDuringRender, nextValue)) {
+        // Force a re-render.
+        inst.value = nextValue;
+        setVersion(bumpVersion);
+      }
+    } catch (error) {
```

Comment:
What errors do we expect here? Is it proper error or also errors that happen due to tearing that can be fixed? Should this intentionally silence errors or no?

Reply:
Since we also schedule an update, it will throw again during render so you can catch it with an error boundary. Same thing that we do for reducers.